### PR TITLE
feat: Enhancements to namespace

### PR DESCRIPTION
**Pull request by JigarJoshi**
Enhanced namespace from string:uint32 pair to structured object.

### CODE REVIEW

File: server/metadata/dictionary.go [hunk 1]
```diff
@@ -133,16 +144,30 @@ func (r *reservedSubspace) reload(ctx context.Context, tx transaction.Tx) error
...
+				Id:          allocatedValue,
+				Name:        allocatedTo.(string),
+				DisplayName: allocatedTo.(string), // for legacy use display name same as the namespace name
+			}
+			r.idToNamespaceStruct[allocatedValue] = namespaceMetadata
+			r.namespaceToNamespaceStruct[allocatedTo.(string)] = namespaceMetadata
+		} else if row.Data.Encoding == namespace_json_encoding {
+			var allocatedValue NamespaceMetadata
+			err = json.Unmarshal(row.Data.RawData, &allocatedValue)
+			if err != nil {
```
Comment (himank) [inline server/metadata/dictionary.go:159]: nit: merge it with above.

File: server/metadata/dictionary.go [hunk 2]
```diff
@@ -153,23 +178,27 @@ func (r *reservedSubspace) reserveNamespace(ctx context.Context, tx transaction.
 
 	r.RLock()
 	defer r.RUnlock()
-	if allocatedTo, ok := r.allocated[id]; ok {
-		log.Debug().Uint32("namespace_id", id).Str("namespace_name", r.allocated[id]).Msg("namespace reserved for")
-		if allocatedTo == namespace {
-			return nil
-		} else {
-			return errors.AlreadyExists("id is already assigned to the namespace '%s'", allocatedTo)
+	if _, ok := r.idToNamespaceStruct[namespaceMetadata.Id]; ok {
```
Comment (himank) [inline server/metadata/dictionary.go:181]: add a `namespaceMetadata.Id` zero check above this as well i.e. if the `id` is zero and passed to this method then we reject.

File: server/metadata/dictionary.go [hunk 3]
```diff
@@ -133,16 +144,30 @@ func (r *reservedSubspace) reload(ctx context.Context, tx transaction.Tx) error
...
 		}
-
-		allocatedValue := ByteToUInt32(row.Data.RawData)
-		r.allocated[allocatedValue] = allocatedTo.(string)
-		r.namespaceToId[allocatedTo.(string)] = allocatedValue
+		if row.Data.Encoding == namespace_int_encoding {
+			allocatedValue := ByteToUInt32(row.Data.RawData)
+			namespaceMetadata := NamespaceMetadata{
+				Id:          allocatedValue,
+				Name:        allocatedTo.(string),
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/metadata/dictionary.go]: fixed

File: server/metadata/dictionary.go (hunk 1 above)
Comment (JigarJoshi) [reply server/metadata/dictionary.go:156]: fixed

File: server/services/v1/management.go [hunk 4]
```diff
@@ -67,21 +64,30 @@ func newManagementService(authProvider AuthProvider, txMgr *transaction.Manager,
...
 	if err != nil {
 		return nil, errors.Internal("Failed to create namespace")
 	}
+
+	if id == 0 {
+		namespaces, err := m.TenantManager.ListNamespaces(ctx, tx)
+		if err != nil {
+			return nil, errors.Internal("Failed to generate ID")
+		}
+		id = uint32(len(namespaces) + 1)
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/management.go]: good point - fixed.

File: server/metadata/tenant.go [hunk 5]
```diff
@@ -48,6 +48,17 @@ type Namespace interface {
 	Id() uint32
 	// Name is the name used for the lookup.
 	Name() string
+	// Metadata for the namespace
+	Metadata() NamespaceMetadata
+}
+
+type NamespaceMetadata struct {
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/metadata/tenant.go:56]: fixed.

File: server/metadata/dictionary.go (hunk 2 above)
Comment (JigarJoshi) [reply server/metadata/dictionary.go:182]: fixed.

**Review by himank — APPROVED** [review]

**Pull request merged**